fish-rust/src/builtins/string/replace.rs

```diff
@@ -0,0 +1,255 @@
+                        cmd,
+                        e.error_message()
+                    );
+                    return STATUS_INVALID_ARGS;
+                }
+            };
+            replace_count += replaced as usize;
+
+            if !self.quiet && (!self.filter || replaced) {
+                streams.out.append(result);
+                if want_newline {
+                    streams.out.append1('\n');
+                }
+            }
+
+            if self.quiet && replace_count > 0 {
+                return STATUS_CMD_OK;
+            }
+        }
+
+        if replace_count > 0 {
+            STATUS_CMD_OK
+        } else {
+            STATUS_CMD_ERROR
+        }
+    }
+}
+
+enum StringReplacer<'args, 'opts> {
+    Regex {
+        replacement: WString,
+        regex: Box<Regex>,
+        opts: &'opts Replace<'args>,
+    },
+    Literal {
+        pattern: Cow<'args, wstr>,
+        replacement: Cow<'args, wstr>,
+        opts: &'opts Replace<'args>,
+    },
+}
+
+impl<'args, 'opts> StringReplacer<'args, 'opts> {
+    fn interpret_escape(arg: &'args wstr) -> Option<WString> {
+        use crate::common::read_unquoted_escape;
+
+        let mut result: WString = WString::with_capacity(arg.len());
+        let mut cursor = arg;
+        while !cursor.is_empty() {
+            if cursor.char_at(0) == '\\' {
+                if let Some(escape_len) = read_unquoted_escape(cursor, &mut result, true, false) {
+                    cursor = cursor.slice_from(escape_len);
+                } else {
+                    // invalid escape
+                    return None;
+                }
+            } else {
+                result.push(cursor.char_at(0));
+                cursor = cursor.slice_from(1);
+            }
+        }
+        return Some(result);
+    }
+
+    fn new(
+        pattern: &'args wstr,
+        replacement: &'args wstr,
+        opts: &'opts Replace<'args>,
+    ) -> Result<Self, RegexError> {
+        let r = match (opts.regex, opts.ignore_case) {
+            (true, _) => {
+                let regex = RegexBuilder::new()
+                    .caseless(opts.ignore_case)
+                    .never_utf(true)
+                    .build(pattern.as_char_slice())
+                    .map_err(|e| RegexError::Compile(pattern.to_owned(), e))?;
+
+                let replacement = if feature_test(FeatureFlag::string_replace_backslash) {
+                    replacement.to_owned()
+                } else {
+                    Self::interpret_escape(replacement)
+                        .ok_or_else(|| RegexError::InvalidEscape(pattern.to_owned()))?
+                };
+                Self::Regex {
+                    replacement,
+                    regex: Box::new(regex),
+                    opts,
+                }
+            }
+            (false, true) => Self::Literal {
+                pattern: Cow::Owned(pattern.to_lowercase()),
```

Comment:
As I mentioned earlier, I think it would both end up both with less code and be less tricky for us if we used the `literal` option on our regex, but will leave a comment